libr/bin/format/elf/elf.c

```diff
@@ -1180,6 +1180,186 @@ struct r_bin_elf_section_t* Elf_(r_bin_elf_get_sections)(struct Elf_(r_bin_elf_o
 	return ret;
 }
 
+static struct r_bin_elf_symbol_t* get_symbols_from_phdr (struct Elf_(r_bin_elf_obj_t) *bin, int type) {
+	Elf_(Sym) *sym = NULL;
+	Elf_(Dyn) *dyn_buf = NULL;
+	Elf_(Addr) addr_sym_table, strtabaddr = 0;
+	char *strtab = NULL;
+	size_t strsize = 0;
+	struct r_bin_elf_symbol_t *ret = NULL;
+	int entries;
+	int i, j, k, r, tsize, len, nsym, ret_ctr;
+	ut64 toffset;
+
+	if (!bin || !bin->phdr || bin->ehdr.e_phnum == 0)
+		return NULL;
+
+	for (i = 0; i < bin->ehdr.e_phnum ; i++){
+		if (bin->phdr[i].p_type == PT_DYNAMIC) break;
+	}
+	if (i == bin->ehdr.e_phnum){
+		// we didn't find the PT_DYNAMIC section
+		return NULL;
+	}
+	entries = bin->phdr[i].p_filesz / sizeof (Elf_(Dyn));
+	dyn_buf = (Elf_(Dyn)*)calloc (entries, sizeof (Elf_(Dyn)));
+	if (!dyn_buf) return NULL;
+	r = r_buf_read_at (bin->b, bin->phdr[i].p_offset, (ut8 *)dyn_buf, bin->phdr[i].p_filesz);
+	if (r == -1 || r == 0){
+		free (dyn_buf);
+		return NULL;
+	}
+	for (j = 0; j < entries; j++) {
+		if (dyn_buf[j].d_tag == DT_SYMTAB) {
+			addr_sym_table = dyn_buf[j].d_un.d_ptr - bin->baddr;
+		}
+		if (dyn_buf[j].d_tag == DT_STRTAB) {
+			strtabaddr = dyn_buf[j].d_un.d_ptr - bin->baddr;
+		}
+		if (dyn_buf[j].d_tag == DT_STRSZ) {
+			strsize = dyn_buf[j].d_un.d_val;
+		}
+	}
+	if (!strtabaddr || strsize > ST32_MAX || strsize == 0){
+		free (dyn_buf);
+		return NULL;
+	}
+	//since  bin doesn't have it let write that information ??
+	strtab = (char *)calloc(1, strsize);
+	if (!strtab){
+		free (dyn_buf);
+		return NULL;
+	}
+	r = r_buf_read_at (bin->b, strtabaddr, (ut8 *)strtab, strsize);
+	if (r == 0 || r == -1){
+		free (strtab);
+		free (dyn_buf);
+		return NULL;
+	}
+
+	if (addr_sym_table){
+		//since ELF doesn't specify the symbol table size we are going to read until the end of the buffer
+		// this might be overkill. 
+		nsym = (bin->b->length - addr_sym_table) / sizeof (Elf_(Sym)); 
+		sym = (Elf_(Sym)*) calloc (nsym, sizeof (Elf_(Sym)));
+		if (!sym){
+			free (dyn_buf);
+			free (strtab);
+			return NULL;
+		}
+		r = r_buf_fread_at (bin->b, addr_sym_table , (ut8*)sym,
+#if R_BIN_ELF64
```

Comment:
It was done by nibble. In rbuf. Just read the src



> On 24 Apr 2015, at 21:36, jvoisin <notifications@github.com> wrote:
> 
> In libr/bin/format/elf/elf.c:
> 
> > +		free (dyn_buf);
> > +		return NULL;
> > +	}
> > +
> > +	if (addr_sym_table){
> > +		//since ELF doesn't specify the symbol table size we are going to read until the end of the buffer
> > +		// this might be overkill. 
> > +		nsym = (bin->b->length - addr_sym_table) / sizeof (Elf_(Sym)); 
> > +		sym = (Elf_(Sym)*) calloc (nsym, sizeof (Elf_(Sym)));
> > +		if (!sym){
> > +			free (dyn_buf);
> > +			free (strtab);
> > +			return NULL;
> > +		}
> > +		r = r_buf_fread_at (bin->b, addr_sym_table , (ut8*)sym,
> > +#if R_BIN_ELF64
> Do you have some link to resources about this? Except in radare2, I have never seen this syntax.
> 
> —
> Reply to this email directly or view it on GitHub.
>